### ISSUE TITLE
Wire up all optimization in a good order

### ISSUE THREAD

**Issue by VMatthijs** [stan-dev/stanc3#286]
As noted by @rybern , it is a hard problem (known as the phase ordering problem) to decide on the optimal order to apply optimization passes. Sometimes, it can even be useful to apply the same optimization multiple times.

The point of this issue is to come up with a sensible order of applying the optimization passes in stanc3. This should be backed up both by common sense and by some empirics.

Some first thoughts on constraints we want on the order (where < means "is performed before"):

function_inlining < basically everything, as most optimizations do not work yet in function bodies

one_step_loop_unrolling < lazy_code_motion, to get loop-invariant code motion

constant_propagation < static_loop_unrolling, to create opportunities for unrolling statically-sized loops

list_collapsing should perhaps happen after some optimizations if they introduce redundant nesting of SList-structures

everything < block_fixing, to make sure that SList constructors directly under if, for, while or fundef constructors are replaced with Block constructors, to make sure the C++ compiles

expression_propagation < partial_evaluation, to create more opportunities for algebraic simplification

partial_evaluation < lazy_code_motion, to reduce subcomputations to normal forms before we get rid of repeated computation (meaning that more subcomputations get shared)

lazy_code_motion < copy_propagation, to clean up (probably, check that this is necessary though)

basically everything < dead_code_elimination, as that will lead to the most aggressive dead-code elimination

basically everything < optimize_ad_levels, as that will lead to the most optimal AD-levels

**Comment by rybern** [comment]
These rules seems like a good starting point. The order in the optimization-integration branch [here](https://github.com/stan-dev/stanc3/blob/7b89b173e3456c3b6a0892ad85c2576ef2ca514c/src/analysis_and_optimization/Optimize.ml#L880) reflects these now, I think.

It might be a fun self-contained project for someone to optimize the phase order with empirical measurements.

**Comment by rybern** [comment]
Solved by #292
